### ISSUE TITLE
Mechanism for foldable aggregations in ES|QL

### ISSUE THREAD

**Issue by astefan** [elastic/elasticsearch#118292]
### Description

Currently, ES|QL lacks two major performance improvements aspects for aggregations (`stats`):
- [planner (LogicalPlanOptimizer component) propagation of foldable expressions into aggregations](https://github.com/elastic/elasticsearch/issues/100634). This is mainly about extending the functionality in `PropagateEvalFoldables` optimization rule.
- short-circuiting `null` handling in aggregation functions so that aggregations on `null` constant do not reach the compute engine and is already computed locally. There are multiple [already](https://github.com/elastic/elasticsearch/issues/110257) [created](https://github.com/elastic/elasticsearch/issues/104430) [issues](https://github.com/elastic/elasticsearch/issues/100170) about this.

[A first attempt](https://github.com/elastic/elasticsearch/pull/112392) at addressing this introduced too many workarounds to be considered valuable and merge-worthy:
- [Calling `SubstituteSurrogates` twice](https://github.com/elastic/elasticsearch/pull/112392/files#diff-45f172b501f6b73039f2a486c29da73ed84d03468540659d86dfa8e10222ade9R170). The `SubstituteSurrogates` rule should be called only once, its role shouldn't be a partial mechanism for dealing with constants in aggregate functions, as it happens now.
- surrogate expression replacement before the entire set of constants have been folded and propagated can introduce issues for validation of arguments correctness. See [this situation](https://github.com/elastic/elasticsearch/pull/112392/files#r1802734728) where currently this is an issue. More about this concept [here](https://github.com/elastic/elasticsearch/pull/112392/files#r1738490217) and [here](https://github.com/elastic/elasticsearch/pull/112392/files#r1738501593) as well.
- having `SubstituteSurrogates` be called twice introduced yet another issue and [broke existent working behavior](https://github.com/elastic/elasticsearch/pull/112392/files#r1861727143)

As a consequence, aggregations must support a specific mechanism in which they can provide a result given they receive constants as arguments, probably in the form of an interface. This should also come with support in the LogicalPlanOptimizer in the form of a new rule + changes to existent rules that deal with folding, null folding and constants propagation.
